dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/PersistablesFilterTest.java

```diff
@@ -0,0 +1,821 @@
+                () -> assertContainsOnly( persistable, Event.class, "Qck4PQ7TMun", "Ox1qBWsnVwE", "jNyGqnwryNi" ),
+                () -> assertContainsOnly( persistable, Relationship.class, "Te3IC6TpnBB"),
+                () -> assertIsEmpty( persistable.getErrors())
+        );
+    }
+
+    @Test
+    void testDeleteInvalidTrackedEntityCannotBeDeletedButItsChildrenCan()
+    {
+        // @formatter:off
+        Setup setup = new Setup.Builder()
+                .trackedEntity( "xK7H53f4Hc2" ).isNotValid()
+                    .enrollment( "t1zaUjKgT3p" )
+                        .event( "Qck4PQ7TMun" )
+                        .event( "Ox1qBWsnVwE" )
+                .build();
+
+        PersistablesFilter.Result persistable = filter( setup.bundle, setup.invalidEntities,
+                TrackerImportStrategy.DELETE );
+
+        assertAll(
+                () -> assertIsEmpty(persistable.get(TrackedEntity.class)),
+                () -> assertContainsOnly( persistable, Enrollment.class, "t1zaUjKgT3p"),
+                () -> assertContainsOnly( persistable, Event.class, "Qck4PQ7TMun", "Ox1qBWsnVwE"),
+                () -> assertIsEmpty( persistable.getErrors())
+        );
+    }
+
+    @Test
+    void testDeleteOnlyReportErrorsIfItAddsNewInformation()
+    {
+        // If entities are found to be invalid during the validation an error for the entity will already be in the
+        // validation report. Only add errors if it would not be clear why an entity cannot be persisted.
+
+        // @formatter:off
+        Setup setup = new Setup.Builder()
+                .trackedEntity( "xK7H53f4Hc2" ).isNotValid()
+                    .enrollment( "t1zaUjKgT3p" ).isNotValid()
+                .relationship("Te3IC6TpnBB",
+                        trackedEntity("xK7H53f4Hc2"),
+                        enrollment("t1zaUjKgT3p") ).isNotValid()
+                .build();
+
+        PersistablesFilter.Result persistable = filter( setup.bundle, setup.invalidEntities,
+                TrackerImportStrategy.DELETE );
+
+        assertAll(
+                () -> assertIsEmpty(persistable.get(TrackedEntity.class)),
+                () -> assertIsEmpty(persistable.get(Enrollment.class)),
+                () -> assertIsEmpty(persistable.get(Relationship.class)),
+                () -> assertIsEmpty( persistable.getErrors())
+        );
+    }
+
+    @Test
+    void testDeleteInvalidRelationshipPreventsDeletionOfTrackedEntityAndEvent()
+    {
+        // @formatter:off
+        Setup setup = new Setup.Builder()
+                .trackedEntity( "xK7H53f4Hc2" )
+                    .enrollment( "t1zaUjKgT3p" )
+                        .event( "Qck4PQ7TMun" )
+                .trackedEntity( "QxGbKYwChDM" )
+                    .enrollment( "Ok4Fe5moc3N" )
+                .relationship("Te3IC6TpnBB",
+                        trackedEntity("xK7H53f4Hc2"),
+                        event("Qck4PQ7TMun") ).isNotValid()
```

Comment:
Just small suggestion, maybe it would be more readable if the `.isNotValid()` goes on the next line, aligned with relationships?

Just suggesting because it is the second time I mistakenly think the event is invalid

Reply:
Will do 👌